README.md
# Thermocrystal1D


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2017-07-29
### Added
- Build with a Makefile

### Changed
- Amplitude for a running wave is now a half of the standing wave's amplitude
- Change filename for a standing wave's template

Sources/main.cc
#include <iostream>
#include "mpi.h"
#include "ai.hh"

#include "thermocrystal1D.hh"

int main(int argc, char *argv[]){
    bool standingWave = false;
    bool runningWave = false;
    bool saveVelocities = false;
    bool runningFromCLI = false;
    
    size_t id = 0;
    size_t length = 1000;
    
    double dispersion = 1.;
    double modelingTime = 100000;
    
    std::string path("./");    
    std::string filename("");
    
    for(int i = 1; i < argc; ++i){
        if("-h" == std::string(argv[i]) || "--help" == std::string(argv[i])){
            std::cout << "usage: ./thermocrystal1D [options]"
                << std::endl
                << "    -h  --help            print this usage and exit"
                << std::endl<< std::endl
                
                << "    --path=<path>         path to save files [string]"
                << std::endl
                << "    --filename=<path>     name of the output file [string]"
                << std::endl
                << "    --id=<value>          positive identification number "
                << "[size_t]"
                << std::endl
                << "    --length=<value>      length of the crystal in "
                << "particles [size_t]"
                << std::endl
                << "    --dispersion=<value>  ratio of thermal energy to "
                << "mechanic energy [double]"
                << std::endl
                << "    --time=<value>        modeling time [double]"
                << std::endl
                << "    --save-velocities     flag to save velocities [bool]"
                << std::endl
                << "    --standing-wave       flag to calculate a stanging "
                << "wave [bool]"
                << std::endl
                << "    --running-wave        flag to calculate a running "
                << "wave [bool]"
                << std::endl
                << "    --env=cli             flag for CLI execution [bool]"
                << std::endl;
            
            return 0;
        }
        
        if(
            ai::assignStringParameter(argv[i], "--path=", path)
            || ai::assignStringParameter(argv[i], "--filename=", filename)
            || ai::assignParameter(argv[i], "--id=", id)
            || ai::assignParameter(argv[i], "--length=", length)
            || ai::assignParameter(argv[i], "--dispersion=", dispersion)
            || ai::assignParameter(argv[i], "--time=", modelingTime)
            || ai::assignBooleanParameter(argv[i], "--save-velocities", 
                saveVelocities)
            || ai::assignBooleanParameter(argv[i], "--standing-wave", 
                standingWave)
            || ai::assignBooleanParameter(argv[i], "--running-wave", 
                runningWave)
            || ai::assignBooleanParameter(argv[i], "--env=cli", 
                runningFromCLI)
        ){
            continue;
        }
    }
    
    if(!standingWave && !runningWave){
        if(!runningFromCLI){
            std::cout << "No wave has been specified. Go with a running wave."
                << std::endl;
        }
        runningWave = true;
    }
    
    if(!ai::hasSuffix(path, "/")){
        path += std::string("/");
    }
    
    const std::string folder(path);
    
    path += filename;
    
    MPI::Init();
    
    const size_t rank = (size_t) MPI::COMM_WORLD.Get_rank();
    const size_t size = (size_t) MPI::COMM_WORLD.Get_size();
    
    if(runningWave){
        thermocrystal1D(
            path + ai::string("rw_") + ai::string(id),
            std::string("running"),
            dispersion,
            modelingTime,
            length,
            saveVelocities,
            0.05,
            runningFromCLI
        );
    }
    
    MPI::COMM_WORLD.Barrier();
    
    if(standingWave){
        thermocrystal1D(
            path + ai::string("sw_") + ai::string(id),
            std::string("standing"),
            dispersion,
            modelingTime,
            length,
            saveVelocities,
            0.05,
            runningFromCLI
        );
    }
    
    if(0 == rank){
        ai::saveVector(
            folder + ai::string("output") + ai::string(id),
            std::vector<double>{(double) length, dispersion, (double) size},
            std::string("#Length; dispersion; cluster size")
        );
    }
    
    MPI::Finalize();
    
    return 0;
}


Sources/thermocrystal1D.cc
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <random>
#include "mpi.h"
#include "ai.hh"

void applyPeriodic(std::vector<double> &quantity){
    const size_t length = quantity.size();
    
    quantity[0] = quantity[length - 2];
    quantity[length - 1] = quantity[1];
}

/*/
 * dispersion is a ratio of thermal energy to mechanic energy
 * set waveType either to "running" or to "standing"
/*/
void thermocrystal1D(
    const std::string filename,
    const std::string waveType,
    const double dispersion = 1.,
    const double maximumTime = 100000,
    size_t length = 1000,
    const bool saveVelocities = false,
    const double dt = 0.05,
    const bool runningFromCLI = false
){
    
    bool runningWaveFlag = false;
    
    const size_t stepToSaveData = 3000;
    const size_t precision = 100000000;
    const size_t rank = (size_t) MPI::COMM_WORLD.Get_rank();
    const size_t size = (size_t) MPI::COMM_WORLD.Get_size();
    
    const double dx = 1.;
    const double alpha = 1.;
    double amplitude = 0.02;
    const double omegaSquare = 1.;
    
    if(!runningFromCLI){
        std::cout << "Incoming parameters:" << std::endl
            << waveType << " wave with " << length << " particles," << std::endl
            << "dispersion is " << dispersion << ", time is " << maximumTime 
            << std::endl << "cluster size is " << size << std::endl;
    }
    
    std::vector<double> displacement(length + 2, 0.);
    std::vector<double> velocity(length + 2, 0.);
    
    if(std::string("running") == waveType){
        runningWaveFlag = true;
        /*/ running wave goes both way /*/
        /*/ we calculate propagation only in one direction /*/
        amplitude = 0.01;
    }
    
    size_t step = 0;
    
    double currentTime = 0.;
    /*/ need in case of a running wave only /*/
    const double shiftTime = dx / sqrt(omegaSquare);
    double savedTime = 0.;
    
    double *arrayBuffer;
    double *velocityArray;
    
    double valueBuffer = 0;
    
    std::vector<double> energyOutput;
    
    std::vector< std::vector<double> > velocityOutput;
    
    if(saveVelocities){
        velocityArray = new double[length];
        
        if(0 == rank){
            arrayBuffer = new double[size * length];
        }
    }
    
    std::random_device rd;
    srand(rd() * time(0));
    
    /*/ now applying initial conditions /*/
    double sum = 0;
    
    /*/ thermal velocities /*/
    for(size_t i = 1; i <= length; ++i){
        velocity[i] = amplitude * sqrt(dispersion * 3. / 2.) 
            * (2. / precision) * ((rand() % precision) - 1.);
        sum += velocity[i];
    }
    
    sum /= (double) length;
    
    /*/ mechanic wave velocities /*/
    for(size_t i = 1; i <= length; ++i){
        double x = 2. * M_PI * (i - 1.) / ((double) length);
        
        velocity[i] += amplitude * sin(x) - sum;
        
        /*/ for a case of a standing wave displacements are already zero /*/
        if(runningWaveFlag){
            displacement[i] = amplitude * cos(x) 
                * ((double) length) / (2. * M_PI * sqrt(omegaSquare));
        }
    }
    
    /*/ now applying periodic boundary conditions /*/
    applyPeriodic(velocity);
    applyPeriodic(displacement);
    
    while(maximumTime > currentTime){
        /*/ calculating a step /*/
        for(size_t i = 1; i <= length; ++i){
            velocity[i] += (displacement[i + 1] - 2 * displacement[i] 
                + displacement[i - 1]) * (omegaSquare + alpha * (displacement[i + 1] 
                - displacement[i - 1])) * dt / dx / dx;
        }
        
        for(size_t i = 1; i <= length; ++i){
            displacement[i] += velocity[i] * dt;
        }
        
        applyPeriodic(velocity);
        applyPeriodic(displacement);
        
        currentTime += dt;
        ++step;
        
        /*/ save the data /*/
        if(stepToSaveData < step){
            step = 0;
            
            if(saveVelocities){
                /*/ gather the velocities /*/
                for(size_t i = 1; i <= length; ++i){
                    velocityArray[i - 1] = velocity[i];
                }
                
                MPI::COMM_WORLD.Reduce(velocityArray, arrayBuffer, 
                    (int) length, MPI::DOUBLE, MPI::SUM, 0);
                
                if(0 == rank){
                    std::vector<double> velocityBuffer;
                    
                    for(size_t i = 0; i < length; ++i){
                        velocityBuffer.push_back(arrayBuffer[i]);
                    }
                    
                    velocityOutput.push_back(velocityBuffer);
                }
            }
            
            /*/ calculating energy /*/
            double kineticEnergy = 0;
            double potentialEnergy = 0;
            
            for(int i = 1; i <= length; ++i){
                float x = (i - 1.) / ((double) length);
                
                kineticEnergy += velocity[i] * sin(2 * M_PI * x);
                potentialEnergy += (displacement[i - 1] - displacement[i]) 
                    * cos(2 * M_PI *x);
            }
            kineticEnergy *= kineticEnergy;
            potentialEnergy *= potentialEnergy;
            
            double energy = potentialEnergy + kineticEnergy;
            
            /*/ gather the energy /*/
            MPI::COMM_WORLD.Reduce(&energy, &valueBuffer, 1, MPI::DOUBLE, MPI::SUM, 0);
            
            if(0 == rank){
                energyOutput.push_back(valueBuffer);
                
                if(!runningFromCLI){
                    ai::showProgressBar(currentTime / maximumTime);
                }
            }
        }
        
        /*/ in cave of a running wave move the wave forward /*/
        if(runningWaveFlag && shiftTime < currentTime - savedTime){
            for(size_t i = 1; i <= length; ++i){
                displacement[i] = displacement[i + 1];
                velocity[i] = velocity[i + 1];
            }
            
            applyPeriodic(velocity);
            applyPeriodic(displacement);
            
            savedTime += shiftTime;
        }
    }
    
    if(0 == rank && !runningFromCLI){
        ai::showProgressBar(1.);
    }
    
    /*/ saving the data /*/
    if(0 == rank){
        if(saveVelocities){
            ai::saveMatrix(filename, velocityOutput);
        }
        
        ai::saveVector(filename, energyOutput);
    }
    
    if(saveVelocities){
        delete [] velocityArray;
        
        if(0 == rank){
            delete [] arrayBuffer;
        }
    }
}
